File: VB/RepositoryItemCustomFontEdit.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Drawing

Namespace CustomFontEditControl
	Friend Class RepositoryItemCustomFontEdit
		Inherits RepositoryItemFontEdit
		' static constructor which calls static registration method
		Shared Sub New()
			RegisterCustomFontEdit()
		End Sub

		' static register method
		Public Shared Sub RegisterCustomFontEdit()
            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(CustomFontEditorName, GetType(CustomFontEdit), GetType(RepositoryItemCustomFontEdit), GetType(ComboBoxViewInfo), New ButtonEditPainter(), True))
		End Sub

		Friend Shared Sub DrawFontName(ByVal gr As Graphics, ByVal name As String, ByVal font As Font, ByVal normalFont As Font, ByVal brush As Brush, ByVal bounds As Rectangle, ByVal showPreview As Boolean)
			Using sf As New StringFormat()
				sf.FormatFlags = sf.FormatFlags Or StringFormatFlags.NoWrap
				If font.Height < bounds.Height OrElse font.Height \ bounds.Height > 2 Then
					sf.LineAlignment = StringAlignment.Center
				End If
				If (Not DevExpress.Utils.ControlUtils.IsSymbolFont(font)) Then
					Try
						gr.DrawString(name, font, brush, bounds, sf)
					Catch
						gr.DrawString(name, normalFont, brush, bounds, sf)
					End Try
				Else
					gr.DrawString(name, normalFont, brush, bounds)
					If showPreview Then
						Dim size As SizeF = gr.MeasureString(name & "w", normalFont)
						Dim newBounds As RectangleF = bounds
						newBounds.Offset(size.Width, 0)
						gr.DrawString(name, font, brush, newBounds, sf)
					End If
				End If
			End Using
		End Sub

		Friend Shared Overloads Sub DrawFontBoxItem(ByVal e As ListBoxDrawItemEventArgs, ByVal item As RepositoryItemCustomFontEdit, ByVal IsDrawStyle As Boolean, ByVal sCurrentFontName As String)
			If sCurrentFontName = String.Empty Then
				Return
			End If
			Using foreBrush As New SolidBrush(e.Appearance.ForeColor), backBrush As New SolidBrush(e.Appearance.BackColor)
				If (Not FontItemPaintHelper.DrawItemBar(e, TryCast(item, RepositoryItemComboBox))) Then
					e.Graphics.FillRectangle(backBrush, e.Bounds)
				End If
				Using family As New FontFamily(sCurrentFontName)
					Dim currentStyle As FontStyle = FontStyle.Regular
					If IsDrawStyle Then
						Select Case e.Item.ToString()
							Case "Regular"
								currentStyle = FontStyle.Regular
							Case "Italic"
								currentStyle = FontStyle.Italic
							Case "Bold"
								currentStyle = FontStyle.Bold
							Case "Bold Italic"
								currentStyle = FontStyle.Bold Or FontStyle.Italic
							Case Else
						End Select
					End If
					Using font As New Font(family, e.Appearance.Font.Size, currentStyle)
						RepositoryItemCustomFontEdit.DrawFontName(e.Graphics, e.Item.ToString(), font, e.Appearance.Font, foreBrush, e.Bounds, True)
						e.Handled = True
					End Using
				End Using
			End Using
		End Sub

		' internal editor name
		Friend Const CustomFontEditorName As String = "CustomFontEdit"

		' public constructor
		Public Sub New()
			MyBase.New()
		End Sub

		' ovverride property
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return CustomFontEditorName
			End Get
		End Property

		Public Overrides Overloads Function GetDisplayText(ByVal format As DevExpress.Utils.FormatInfo, ByVal editValue As Object) As String
			If editValue IsNot Nothing AndAlso (TryCast(editValue, Font)) IsNot Nothing Then
				Return String.Format("{0}, {1}", (TryCast(editValue, Font)).Name, (TryCast(editValue, Font)).Size)
			Else
				Return MyBase.GetDisplayText(format, editValue)
			End If
		End Function
	End Class
End Namespace

File: CS/RepositoryItemCustomFontEdit.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors;
using System.ComponentModel;
using System.Windows.Forms;
using System.Drawing;

namespace CustomFontEditControl
{
    internal class RepositoryItemCustomFontEdit : RepositoryItemFontEdit
    {
        // static constructor which calls static registration method
        static RepositoryItemCustomFontEdit() { RegisterCustomFontEdit(); }

        // static register method
        public static void RegisterCustomFontEdit()
        {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(
                CustomFontEditorName,
                typeof(CustomFontEdit),
                typeof(RepositoryItemCustomFontEdit),
                typeof(ComboBoxViewInfo),
                new ButtonEditPainter(),
                true,
                null));
        }

        internal static void DrawFontName(Graphics gr, string name, Font font, Font normalFont, Brush brush, Rectangle bounds, bool showPreview)
        {
            using (StringFormat sf = new StringFormat())
            {
                sf.FormatFlags |= StringFormatFlags.NoWrap;
                if (font.Height < bounds.Height || font.Height / bounds.Height > 2)
                    sf.LineAlignment = StringAlignment.Center;
                if (!DevExpress.Utils.ControlUtils.IsSymbolFont(font))
                {
                    try
                    {
                        gr.DrawString(name, font, brush, bounds, sf);
                    }
                    catch
                    {
                        gr.DrawString(name, normalFont, brush, bounds, sf);
                    }
                }
                else
                {
                    gr.DrawString(name, normalFont, brush, bounds);
                    if (showPreview)
                    {
                        SizeF size = gr.MeasureString(name + "w", normalFont);
                        RectangleF newBounds = bounds;
                        newBounds.Offset(size.Width, 0);
                        gr.DrawString(name, font, brush, newBounds, sf);
                    }
                }
            }
        }

        internal static void DrawFontBoxItem(ListBoxDrawItemEventArgs e, RepositoryItemCustomFontEdit item, bool IsDrawStyle, string sCurrentFontName)
        {
            if (sCurrentFontName == string.Empty) return;
            using (SolidBrush foreBrush = new SolidBrush(e.Appearance.ForeColor), backBrush = new SolidBrush(e.Appearance.BackColor))
            {
                if (!FontItemPaintHelper.DrawItemBar(e, item as RepositoryItemComboBox))
                    e.Graphics.FillRectangle(backBrush, e.Bounds);
                using (FontFamily family = new FontFamily(sCurrentFontName))
                {
                    FontStyle currentStyle = FontStyle.Regular;
                    if (IsDrawStyle)
                    {
                        switch (e.Item.ToString())
                        {
                            case "Regular":
                                currentStyle = FontStyle.Regular;
                                break;
                            case "Italic":
                                currentStyle = FontStyle.Italic;
                                break;
                            case "Bold":
                                currentStyle = FontStyle.Bold;
                                break;
                            case "Bold Italic":
                                currentStyle = FontStyle.Bold | FontStyle.Italic;
                                break;
                            default:
                                break;
                        }
                    }
                    using (Font font = new Font(family, e.Appearance.Font.Size, currentStyle))
                    {
                        RepositoryItemCustomFontEdit.DrawFontName(e.Graphics, e.Item.ToString(), font, e.Appearance.Font, foreBrush, e.Bounds, true);
                        e.Handled = true;
                    }
                }
            }        
        }

        // internal editor name
        internal const string CustomFontEditorName = "CustomFontEdit";

        // public constructor
        public RepositoryItemCustomFontEdit() : base()  {  }

        // ovverride property
        public override string EditorTypeName
        {
            get { return CustomFontEditorName; }
        }

        public override string GetDisplayText(DevExpress.Utils.FormatInfo format, object editValue)
        {
            if (editValue != null && (editValue as Font) != null)
                return String.Format("{0}, {1}", (editValue as Font).Name, (editValue as Font).Size);
            else
                return base.GetDisplayText(format, editValue);
        }
    }
}
